core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java

```diff
@@ -0,0 +1,434 @@
+    // CH requests and preparation is unaffected as cached weighting from preparation is used.
+    // Use accessOrder==true to remove oldest accessed entry, not oldest inserted.
+    private static final int CACHE_SIZE = Integer.getInteger("graphhopper.custom_weighting.cache_size", 1000);
+    private static final Map<String, Class<?>> CACHE = Collections.synchronizedMap(
+            new LinkedHashMap<String, Class<?>>(CACHE_SIZE, 0.75f, true) {
+                protected boolean removeEldestEntry(Map.Entry eldest) {
+                    return size() > CACHE_SIZE;
+                }
+            });
+
+    // This internal cache ensures that the "internal" Weighting classes specified in the profiles, are never removed regardless
+    // of how frequent other Weightings are created and accessed. We only need to synchronize the get and put methods alone.
+    // E.g. we do not care for the race condition where two identical classes are requested and one of them is overwritten.
+    // TODO perf compare with ConcurrentHashMap, but I guess, if there is a difference at all, it is not big for small maps
+    private static final Map<String, Class<?>> INTERNAL_CACHE = Collections.synchronizedMap(new HashMap<>());
+
+    private CustomModelParser() {
+        // utility class
+    }
+
+    public static CustomWeighting createWeighting(FlagEncoder baseFlagEncoder, EncodedValueLookup lookup, TurnCostProvider turnCostProvider,
+                                                  CustomModel customModel) {
+        if (customModel == null)
+            throw new IllegalStateException("CustomModel cannot be null");
+        CustomWeighting.Parameters parameters = create(customModel, lookup, baseFlagEncoder.getMaxSpeed(), baseFlagEncoder.getAverageSpeedEnc());
+        return new CustomWeighting(baseFlagEncoder, turnCostProvider, parameters);
+    }
+
+    /**
+     * This method compiles a new subclass of CustomWeightingHelper composed from the provided CustomModel caches this
+     * and returns an instance.
+     */
+    static CustomWeighting.Parameters create(CustomModel customModel, EncodedValueLookup lookup, double globalMaxSpeed,
+                                             DecimalEncodedValue avgSpeedEnc) {
+        String key = customModel.toString() + ",global:" + globalMaxSpeed;
+        if (key.length() > 100_000) throw new IllegalArgumentException("Custom Model too big: " + key.length());
+
+        Class<?> clazz = customModel.shouldUseInternalCache() ? INTERNAL_CACHE.get(key) : null;
+        if (CACHE_SIZE > 0 && clazz == null)
+            clazz = CACHE.get(key);
+        if (clazz == null) {
+            clazz = createClazz(customModel, lookup, globalMaxSpeed);
+            if (customModel.shouldUseInternalCache()) {
+                INTERNAL_CACHE.put(key, clazz);
+                if (INTERNAL_CACHE.size() > 100) {
+                    CACHE.putAll(INTERNAL_CACHE);
+                    INTERNAL_CACHE.clear();
```

Comment:
> And why would it grow larger than 100?

It shouldn't, but to avoid memory issues in case of bugs I thought this is the best solution. Not sure.

> So first of all I would rename the CustomModel#cached field to something like CustomModel#permanent/internal/isServerModel

Ah, yes. Internal or serverModel sounds good

> as long as it is not possible to set this flag from the client side

Yes, I do not know how to do this safely :)

Reply:
> It shouldn't, but to avoid memory issues in case of bugs I thought this is the best solution. Not sure.

Hm, but to detect bugs it would be better to throw an exception. 

> Yes, I do not know how to do this safely :)

Well, first of all we can throw an error if it is set to true on a queryCustomModel. But then we can probably also exclude it from Jackson deserialization.

Reply:
> Hm, but to detect bugs it would be better to throw an exception

Ok, yes. You are right. Let us throw an exception!

> Maybe the internal custom models should not be obtained via some cache key in the first place and instead they should be created when the server starts and later be retrieved directly via the profile name for example.

Sounds also good, yes. But for now throw an exception :)

> Well, first of all we can throw an error if it is set to true on a queryCustomModel

Ah, nice. I tried the opposite, enforce server-side to be true, but this was ugly in tests.